### PR TITLE
BLD: `utest` port

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @HaoZeke,

Today I've been working on porting `utest` tests but I'm stuck on one error. When linking `libopenblas.a` to `utest` test executable it was failing with a bunch of `undefined reference to ..._` errors.

I inspected static libraries build with `make` and `meson` with `nm` and I realized that `meson` ones had incorrect names, e.g.:

With `make`:
```
(openblas-dev2) mtsokol@qgpu3:~/OpenBLAS$ nm interface/dsdot.o
0000000000000000 T dsdot_
                 U dsdot_k
(openblas-dev2) mtsokol@qgpu3:~/OpenBLAS$ nm interface/cblas_dsdot.o
0000000000000000 T cblas_dsdot
                 U dsdot_k
```

With `meson`:
```
(openblas-dev2) mtsokol@qgpu3:~/OpenBLAS$ nm interface/dsdot.o
0000000000000000 T dsdot
                 U dsdot_k
(openblas-dev2) mtsokol@qgpu3:~/OpenBLAS$ nm interface/cblas_dsdot.o
0000000000000000 T cblas_dsdot
                 U dsdot_k
```
I found that in `inteface/meson.build` in the 3rd line you set:
```meson
if not build_without_cblas
  interface_args += '-DCBLAS'
endif
```
But `-DCBLAS` should be set to only `'cblas': true` case objects. (`'-DCBLAS'` is still added in line 649 but this is ignored as it's already present in the `compiler_args`). That was the reason why `..._` where missing.

Without that flag I get the same `nm` names as with make (see the diff in this PR).

---

With this fix, now when linking openblas to `utest` executable I only get one error:
```
/home/mtsokol/.conda/envs/openblas-dev2/bin/../lib/gcc/x86_64-conda-linux-gnu/13.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: libopenblas.a(dsdot.c.o): in function `dsdot_':
dsdot.c:(.text+0x19): undefined reference to `dsdot_k'
/home/mtsokol/.conda/envs/openblas-dev2/bin/../lib/gcc/x86_64-conda-linux-gnu/13.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: dsdot.c:(.text+0x30): undefined reference to `dsdot_k'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Weirdly `dsdot_k` is present in the `meson`'s `libopeblas.a` but it's a bit convoluted. There's also `ddsdot_k` there which is incorrect I think is coming from `'base': '?dsdot'` (see the diff).

Once this `dsdot_k` is solved I should have `utest` ported.

**[EDIT]** All fixed now.

### CODE REVIEW

**Comment by mtsokol** [comment]
`dsdot` kernel was missing `-DDSDOT`. Now `openblas_utest` run with `meson test -C build`.

**Comment by mtsokol** [comment]
Ok, it's ready for review! The current version ports the whole `utest` suite to `meson` and modifies/adds `meson.build`'s definitions of: 
- `caxpy`
- `?axpyc`
- `dsdot`
- `?omatcopy_k`
- `?imatcopy_k`
- `?geadd`

File: utest/meson.build [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+openblas_utest_ext_exec = executable(
+    'openblas_utest_ext',
+    sources: sources_utest_ext,
+    link_with: _openblas,
+    dependencies: [dependency('threads')],
+    include_directories: _inc + utest_inc,
+    # Weird strsignal error
+    c_args: _cargs + '-Wno-implicit-function-declaration',
```
Comment (HaoZeke) [inline utest/meson.build]: I'm not sure if you were seeing this too @mtsokol, but for me, even though `utest` clearly includes both `signal` and `string` I was getting an undefined error, so I added this for now.

**Comment by HaoZeke** [comment]
Locally, for me this passes almost everything..

```bash

Summary of Failures:

 5/14 Test REAL Level 2 BLAS             FAIL            0.05s   exit status 1
 6/14 Test REAL Level 3 BLAS             FAIL            0.05s   exit status 1
 7/14 Test DOUBLE PRECISION Level 2 BLAS FAIL            0.05s   exit status 1
 8/14 Test DOUBLE PRECISION Level 3 BLAS FAIL            0.04s   exit status 1
 9/14 Test COMPLEX Level 2 BLAS          FAIL            0.04s   exit status 1
10/14 Test COMPLEX Level 3 BLAS          FAIL            0.03s   exit status 1
11/14 Test COMPLEX*16 Level 2 BLAS       FAIL            0.03s   exit status 1
12/14 Test COMPLEX*16 Level 3 BLAS       FAIL            0.02s   exit status 1
```

File: utest/meson.build (hunk 1 above)
Comment (mtsokol) [reply utest/meson.build]: Right, I got the same warning. I tried adding `-lc` flag but it didn't go away so I ignored it.

**Comment by mtsokol** [comment]
Ok, so those that you listed here are only existing `lapack-netlib/BLAS/TESTING` tests.

Hmm, for me running these only passes (with this PR changes):
```
 1/14 Test REAL Level 1 BLAS                     OK              0.09s
 2/14 Test DOUBLE PRECISION Level 1 BLAS         OK              0.08s
 3/14 Test COMPLEX Level 1 BLAS                  OK              0.07s
 4/14 Test COMPLEX*16 Level 1 BLAS               OK              0.07s
 5/14 Test REAL Level 2 BLAS                     OK              0.06s
 6/14 Test REAL Level 3 BLAS                     OK              0.06s
 7/14 Test DOUBLE PRECISION Level 2 BLAS         OK              0.05s
 8/14 Test DOUBLE PRECISION Level 3 BLAS         OK              0.05s
 9/14 Test COMPLEX Level 2 BLAS                  OK              0.04s
10/14 Test COMPLEX*16 Level 2 BLAS               OK              0.03s
12/14 Test COMPLEX Level 3 BLAS                  OK              0.03s
13/14 Test COMPLEX*16 Level 3 BLAS               OK              0.04s
```

Hmm maybe try running them one more time with `build/` dir removed first.

**Comment by HaoZeke** [comment]
The failures are from missing files, perhaps something to do with paths
```
 5/14 Test REAL Level 2 BLAS                     FAIL            0.05s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2s: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2s < /OpenBLAS/lapack-netlib/BLAS/TESTING/sblat2.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

 6/14 Test REAL Level 3 BLAS                     FAIL            0.05s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3s: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3s < /OpenBLAS/lapack-netlib/BLAS/TESTING/sblat3.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

 7/14 Test DOUBLE PRECISION Level 2 BLAS         FAIL            0.05s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2d: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2d < /OpenBLAS/lapack-netlib/BLAS/TESTING/dblat2.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

 8/14 Test DOUBLE PRECISION Level 3 BLAS         FAIL            0.04s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3d: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3d < /OpenBLAS/lapack-netlib/BLAS/TESTING/dblat3.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

 9/14 Test COMPLEX Level 2 BLAS                  FAIL            0.04s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2c: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2c < /OpenBLAS/lapack-netlib/BLAS/TESTING/cblat2.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

10/14 Test COMPLEX Level 3 BLAS                  FAIL            0.03s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3c: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3c < /OpenBLAS/lapack-netlib/BLAS/TESTING/cblat3.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

11/14 Test COMPLEX*16 Level 2 BLAS               FAIL            0.03s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2z: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat2z < /OpenBLAS/lapack-netlib/BLAS/TESTING/zblat2.in' failed with return code 32512.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

12/14 Test COMPLEX*16 Level 3 BLAS               FAIL            0.02s   exit status 1
―――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――
stderr:
sh: line 1: /OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3z: No such file or directory
Error: Command '/OpenBLAS/bbdir/lapack-netlib/BLAS/TESTING/xblat3z < /OpenBLAS/lapack-netlib/BLAS/TESTING/zblat3.in' failed with return code 32512.
```

**Comment by HaoZeke** [comment]
@mtsokol if you've not run `make clean` the files might be left over from `make` (thereby passing the tests here)

**Comment by mtsokol** [comment]
Let's then maybe comment `subdir('TESTING')` in `lapack-netlib/BLAS/meson.build` and address it when we work on netlib tests?

**Comment by HaoZeke** [comment]
Ah this was closed by mistake (when I merged and deleted `setupIncludes`)

**Pull request closed without merging**